src/core/core.datasetController.js

```diff
@@ -220,53 +297,105 @@ helpers.extend(DatasetController.prototype, {
 		return type && new type({
 			_ctx: me.chart.ctx,
 			_datasetIndex: me.index,
-			_index: index
+			_index: index,
+			_parsed: {}
 		});
 	},
 
+	/**
+	 * @private
+	 */
+	_dataCheck: function() {
+		var me = this;
+		var dataset = me.getDataset();
+		var data = dataset.data || (dataset.data = []);
+
+		// In order to correctly handle data addition/deletion animation (an thus simulate
+		// real-time charts), we need to monitor these data modifications and synchronize
+		// the internal meta data accordingly.
+
+		if (helpers.isObject(data)) {
+			// Object data is currently monitored for replacement only
+			if (me._objectData === data) {
+				return false;
+			}
+			me._data = convertObjectDataToArray(data);
+			me._objectData = data;
+		} else {
+			if (me._data === data && arraysEqual(data, me._dataCopy)) {
+				return false;
+			}
+
+			if (me._data) {
+				// This case happens when the user replaced the data array instance.
+				unlistenArrayEvents(me._data, me);
+			}
+
+			// Store a copy to detect direct modifications.
+			// Note: This is suboptimal, but better than always parsing the data
+			// TODO: Utilize Proxy instead, after IE support is dropped
+			me._dataCopy = data.slice(0);
+
+			if (data && Object.isExtensible(data)) {
+				listenArrayEvents(data, me);
+			}
+			me._data = data;
+		}
+		return true;
+	},
+
+	/**
+	 * @private
+	 */
+	_labelCheck: function() {
+		var me = this;
+		var scale = me._getIndexScale();
+		var labels = scale ? scale._getLabels() : me.chart.data.labels;
+
+		if (me._labels === labels) {
+			return false;
+		}
+
+		me._labels = labels;
+		return true;
+	},
+
 	addElements: function() {
 		var me = this;
-		var meta = me.getMeta();
-		var data = me.getDataset().data || [];
+		var meta = me._cachedMeta;
 		var metaData = meta.data;
-		var i, ilen;
+		var i, ilen, data;
+
+		me._dataCheck();
+		data = me._data;
 
 		for (i = 0, ilen = data.length; i < ilen; ++i) {
 			metaData[i] = metaData[i] || me.createMetaData(i);
 		}
 
 		meta.dataset = meta.dataset || me.createMetaDataset();
+		me._parse(0, ilen);
```

Comment:
My last comment wasn't very clear. What I mean was you can delete this line because `resyncElements` will do it. Right now with this line present, we do the parsing twice.

Reply:
Looks like you are right, removed.